src/log-store/src/kafka/log_store.rs

```diff
@@ -274,3 +279,208 @@ fn check_termination(
+                entries_with_random_data(3, &ctx.entry_builder)
+            } else {
+                // Builds a large entry of size 256KB which is way greater than the configured `max_batch_size` which is 32KB.
+                let large_entry = ctx.entry_builder.with_data([b'1'; 256 * 1024]);
+                vec![large_entry]
+            };
+            // Aggregates entries of all regions.
+            aggregated.push(ctx.expected.clone());
+        }
+        aggregated.into_iter().flatten().collect()
+    }
+
+    /// Reads entries for regions and checks for each region that the gotten entries are identical with the expected ones.
+    async fn check_entries(
+        region_contexts: &HashMap<u64, RegionContext>,
+        logstore: &KafkaLogStore,
+        which: &[u64],
+    ) {
+        for region_id in which {
+            let ctx = &region_contexts[region_id];
+            let stream = logstore
+                .read(&ctx.ns, ctx.flushed_entry_id + 1)
+                .await
+                .unwrap();
+            let got = stream
+                .collect::<Vec<_>>()
+                .await
+                .into_iter()
+                .flat_map(|x| x.unwrap())
+                .collect::<Vec<_>>();
+            assert_eq!(ctx.expected, got);
+        }
+    }
+
+    /// Simulates a flush for regions.
+    fn flush(
+        region_contexts: &mut HashMap<u64, RegionContext>,
+        last_entry_ids: &HashMap<u64, EntryId>,
+    ) {
+        for (region_id, last_entry_id) in last_entry_ids {
+            let ctx = region_contexts.get_mut(region_id).unwrap();
+            ctx.flushed_entry_id = *last_entry_id;
+        }
+    }
+
+    /// Starts a test with:
+    /// * `test_name` - The name of the test.
+    /// * `num_topics` - Number of topics to be created in the preparation phase.
+    /// * `num_regions` - Number of regions involved in the test.
+    /// * `num_appends` - Number of append operations to be performed.
+    /// * `all` - All regions will be involved in an append operation if `all` is true. Otherwise,
+    /// an append operation will only randomly choose a subset of regions.
+    /// * `large` - Builds large entries for each region is `large` is true.
+    async fn test_with(
+        test_name: &str,
+        num_topics: usize,
+        num_regions: usize,
+        num_appends: usize,
+        all: bool,
+        large: bool,
+    ) {
+        let (logstore, topics) = prepare(test_name, num_topics).await;
+        let mut region_contexts = build_contexts(num_regions, &topics);
+        for _ in 0..num_appends {
+            let which = all_or_subset(all, num_regions);
+            let entries = build_entries(&mut region_contexts, &which, large);
+            let last_entry_ids = logstore.append_batch(entries).await.unwrap().last_entry_ids;
+            check_entries(&region_contexts, &logstore, &which).await;
+            flush(&mut region_contexts, &last_entry_ids);
+        }
```

Comment:
How about expanding methods?  Most of them are all used in one place.  It's much easier to understand what the test is doing/asserting. 

```suggestion
        let mut region_contexts = (0..num_regions)
            .map(|i| {
                let topic = &topics[i % topics.len()];
                let ns = new_namespace(topic, i as u64);
                let entry_builder = EntryBuilder::new(ns.clone());
                (
                    i as u64,
                    RegionContext {
                        ns,
                        entry_builder,
                        expected: Vec::new(),
                        flushed_entry_id: 0,
                    },
                )
            })
            .collect();

        for _ in 0..num_appends {
            let which = all_or_subset(all, num_regions);
            let entries = build_entries(&mut region_contexts, &which, large);
            let last_entry_ids = logstore.append_batch(entries).await.unwrap().last_entry_ids;

            for region_id in which {
                let ctx = &region_contexts[&region_id];
                let stream = logstore
                    .read(&ctx.ns, ctx.flushed_entry_id + 1)
                    .await
                    .unwrap();
                let got = stream
                    .collect::<Vec<_>>()
                    .await
                    .into_iter()
                    .flat_map(|x| x.unwrap())
                    .collect::<Vec<_>>();
                assert_eq!(ctx.expected, got);
            }

            for (region_id, last_entry_id) in last_entry_ids {
                let ctx = region_contexts.get_mut(&region_id).unwrap();
                ctx.flushed_entry_id = last_entry_id;
            }
      }
```